webui/react/src/pages/WorkspaceDetails/WorkspaceMembers.tsx

```diff
@@ -0,0 +1,301 @@
+
+  const { canUpdateRoles } = usePermissions();
+
+  const { settings, updateSettings } = useSettings<WorkspaceMembersSettings>(settingsConfig);
+  const userCanAssignRoles = canUpdateRoles({ workspace });
+
+  const mockWorkspaceMembers = useFeature().isOn('mock_workspace_members');
+  
+  knownRoles = mockWorkspaceMembers ? [{
+    id: 1,
+    name: 'Editor',
+    permissions: [],
+  },
+  {
+    id: 2,
+    name: 'Viewer',
+    permissions: [],
+  }] : knownRoles
+
+  const usersAndGroups: UserOrGroup[] = useMemo(
+    () => mockWorkspaceMembers ? [{
+      displayName: 'Test User One Display Name',
+      id: 1,
+      username: 'TestUserOneUserName',
+    },
+    {
+      id: 2,
+      username: 'TestUserTwoUserName',
+    },
+    {
+      groupId: 1,
+      name: 'Test Group 1 Name',
+    },
+    {
+      groupId: 2,
+      name: 'Test Group 2 Name',
+    },
+  ] : [...usersAssignedDirectly, ...groupsAssignedDirectly],
+    [groupsAssignedDirectly, mockWorkspaceMembers, usersAssignedDirectly],
+  );
+
+  const handleNameSearchApply = useCallback(
+    (newSearch: string) => {
+      updateSettings({ name: newSearch || undefined });
+    },
+    [updateSettings],
+  );
+
+  const handleNameSearchReset = useCallback(() => {
+    updateSettings({ name: undefined });
+  }, [updateSettings]);
+
+  const nameFilterSearch = useCallback(
+    (filterProps: FilterDropdownProps) => (
+      <TableFilterSearch
+        {...filterProps}
+        value={settings.name || ''}
+        onReset={handleNameSearchReset}
+        onSearch={handleNameSearchApply}
+      />
+    ),
+    [handleNameSearchApply, handleNameSearchReset, settings.name],
+  );
+
+  const tableSearchIcon = useCallback(() => <Icon name="search" size="tiny" />, []);
+
+  const columns = useMemo(() => {
+    const nameRenderer = (value: string, record: UserOrGroup) => {
+      if (isUser(record)) {
+        const member = record as User;
+        return (
+          <>
+            <div className={css.userAvatarRowItem}>
+              <Avatar size={Size.Medium} userId={member.id} />
+            </div>
+            <div className={css.userRowItem}>
+              {member?.displayName ? (
+                <>
+                  <div>{member.displayName}</div>
+                  <div>{member.username}</div>
+                </>
+              ) : (
+                <div>{member.username}</div>
+              )}
+            </div>
+          </>
+        );
+      }
+      const group = record as V1GroupDetails;
+      return (
+        <>
+          <div className={css.userAvatarRowItem}>
+            <Icon name="group" />
+          </div>
+          <div className={css.userRowItem}>
+            <div>{group.name}</div>
+          </div>
+        </>
+      );
+    };
+
+    const roleRenderer = (value: string, record: UserOrGroup) => {
+      return (
+        <Select
+          className={css.selectContainer}
+          disabled={!userCanAssignRoles}
```

Comment:
Need clarification if a single assignment will be returned from https://github.com/determined-ai/determined-ee/pull/444/files#diff-2507f8d47389cccf4a093c61fdaf6998acb68ef55b5b41476d298617c51d4f71R75 or if I will need to filter the `assignments` returned per group and user to determine the correct role to show in the table.

Reply:
the PR changed; but based on what I said earlier, I believe that you'd get an array / multiple roles, but as long as it's just the pre-canned/preset roles, we can assume there's only zero or one workspace-specific role